### ISSUE TITLE
cannot load such file -- robots (LoadError) with ruby verison 2.5

### ISSUE THREAD

**Issue by bartman081523** [buren/wayback_archiver#51]
> wayback_archiver 
> Traceback (most recent call last):
> 	10: from /usr/local/bin/wayback_archiver:23:in `<main>'
> 	 9: from /usr/local/bin/wayback_archiver:23:in `load'
> 	 8: from /var/lib/gems/2.7.0/gems/wayback_archiver-1.4.0/bin/wayback_archiver:4:in `<top (required)>'
> 	 7: from /usr/lib/ruby/vendor_ruby/rubygems/core_ext/kernel_require.rb:85:in `require'
> 	 6: from /usr/lib/ruby/vendor_ruby/rubygems/core_ext/kernel_require.rb:85:in `require'
> 	 5: from /var/lib/gems/2.7.0/gems/wayback_archiver-1.4.0/lib/wayback_archiver.rb:4:in `<top (required)>'
> 	 4: from /usr/lib/ruby/vendor_ruby/rubygems/core_ext/kernel_require.rb:85:in `require'
> 	 3: from /usr/lib/ruby/vendor_ruby/rubygems/core_ext/kernel_require.rb:85:in `require'
> 	 2: from /var/lib/gems/2.7.0/gems/wayback_archiver-1.4.0/lib/wayback_archiver/url_collector.rb:2:in `<top (required)>'
> 	 1: from /usr/lib/ruby/vendor_ruby/rubygems/core_ext/kernel_require.rb:85:in `require'
> /usr/lib/ruby/vendor_ruby/rubygems/core_ext/kernel_require.rb:85:in `require': cannot load such file -- robots (LoadError)
> 	11: from /usr/local/bin/wayback_archiver:23:in `<main>'
> 	10: from /usr/local/bin/wayback_archiver:23:in `load'
> 	 9: from /var/lib/gems/2.7.0/gems/wayback_archiver-1.4.0/bin/wayback_archiver:4:in `<top (required)>'
> 	 8: from /usr/lib/ruby/vendor_ruby/rubygems/core_ext/kernel_require.rb:85:in `require'
> 	 7: from /usr/lib/ruby/vendor_ruby/rubygems/core_ext/kernel_require.rb:85:in `require'
> 	 6: from /var/lib/gems/2.7.0/gems/wayback_archiver-1.4.0/lib/wayback_archiver.rb:4:in `<top (required)>'
> 	 5: from /usr/lib/ruby/vendor_ruby/rubygems/core_ext/kernel_require.rb:85:in `require'
> 	 4: from /usr/lib/ruby/vendor_ruby/rubygems/core_ext/kernel_require.rb:85:in `require'
> 	 3: from /var/lib/gems/2.7.0/gems/wayback_archiver-1.4.0/lib/wayback_archiver/url_collector.rb:2:in `<top (required)>'
> 	 2: from /usr/lib/ruby/vendor_ruby/rubygems/core_ext/kernel_require.rb:149:in `require'
> 	 1: from /usr/lib/ruby/vendor_ruby/rubygems/core_ext/kernel_require.rb:160:in `rescue in require'
> /usr/lib/ruby/vendor_ruby/rubygems/core_ext/kernel_require.rb:160:in `require': cannot load such file -- robots (LoadError)
> user@tp-ubuntu:~/Downloads/wayback_archiver$ sudo gem install robots
> Successfully installed robots-0.10.1
> Parsing documentation for robots-0.10.1
> Done installing documentation for robots after 0 seconds
> 1 gem installed
> user@tp-ubuntu:~/Downloads/wayback_archiver$ wayback_archiver 
> Traceback (most recent call last):
> 	10: from /usr/local/bin/wayback_archiver:23:in `<main>'
> 	 9: from /usr/local/bin/wayback_archiver:23:in `load'
> 	 8: from /var/lib/gems/2.7.0/gems/wayback_archiver-1.4.0/bin/wayback_archiver:4:in `<top (required)>'
> 	 7: from /usr/lib/ruby/vendor_ruby/rubygems/core_ext/kernel_require.rb:85:in `require'
> 	 6: from /usr/lib/ruby/vendor_ruby/rubygems/core_ext/kernel_require.rb:85:in `require'
> 	 5: from /var/lib/gems/2.7.0/gems/wayback_archiver-1.4.0/lib/wayback_archiver.rb:4:in `<top (required)>'
> 	 4: from /usr/lib/ruby/vendor_ruby/rubygems/core_ext/kernel_require.rb:85:in `require'
> 	 3: from /usr/lib/ruby/vendor_ruby/rubygems/core_ext/kernel_require.rb:85:in `require'
> 	 2: from /var/lib/gems/2.7.0/gems/wayback_archiver-1.4.0/lib/wayback_archiver/url_collector.rb:2:in `<top (required)>'
> 	 1: from /usr/lib/ruby/vendor_ruby/rubygems/core_ext/kernel_require.rb:85:in `require'
> /usr/lib/ruby/vendor_ruby/rubygems/core_ext/kernel_require.rb:85:in `require': cannot load such file -- robots (LoadError)
> 	11: from /usr/local/bin/wayback_archiver:23:in `<main>'
> 	10: from /usr/local/bin/wayback_archiver:23:in `load'
> 	 9: from /var/lib/gems/2.7.0/gems/wayback_archiver-1.4.0/bin/wayback_archiver:4:in `<top (required)>'
> 	 8: from /usr/lib/ruby/vendor_ruby/rubygems/core_ext/kernel_require.rb:85:in `require'
> 	 7: from /usr/lib/ruby/vendor_ruby/rubygems/core_ext/kernel_require.rb:85:in `require'
> 	 6: from /var/lib/gems/2.7.0/gems/wayback_archiver-1.4.0/lib/wayback_archiver.rb:4:in `<top (required)>'
> 	 5: from /usr/lib/ruby/vendor_ruby/rubygems/core_ext/kernel_require.rb:85:in `require'
> 	 4: from /usr/lib/ruby/vendor_ruby/rubygems/core_ext/kernel_require.rb:85:in `require'
> 	 3: from /var/lib/gems/2.7.0/gems/wayback_archiver-1.4.0/lib/wayback_archiver/url_collector.rb:2:in `<top (required)>'
> 	 2: from /usr/lib/ruby/vendor_ruby/rubygems/core_ext/kernel_require.rb:149:in `require'
> 	 1: from /usr/lib/ruby/vendor_ruby/rubygems/core_ext/kernel_require.rb:160:in `rescue in require'
> /usr/lib/ruby/vendor_ruby/rubygems/core_ext/kernel_require.rb:160:in `require': cannot load such file -- robots (LoadError)
> user@tp-ubuntu:~/Downloads/wayback_archiver$ ruby --version
> ruby 2.7.4p191 (2021-07-07 revision a21a3b7d23) [x86_64-linux-gnu]
> 

RESOLUTION: (funtioning, but not valid anymore, see my last comment)
```
sudo apt purge ruby
sudo apt autoremove
sudo apt install curl
sudo apt install git
command curl -sSL https://rvm.io/mpapis.asc | gpg --import -
command curl -sSL https://rvm.io/pkuczynski.asc | gpg --import -
\curl -sSL https://get.rvm.io | bash -s stable
source /home/$USER/.rvm/scripts/rvm
rvm install ruby-3
git clone https://github.com/buren/wayback_archiver
cd wayback_archiver
gem build wayback_archiver.gemspec
gem install wayback_archiver-1.4.0.gem 
wayback_archiver --url www.example.com
echo "source /home/$USER/.rvm/scripts/rvm" >> /home/$USER/.bashrc
```

**Comment by buren** [comment]
@chlorophyll-zz this doesn't seem like this is an issue with `wayback_archiver`?

**Comment by chlorophyll-zz** [comment]
@buren 
I dont know. 
/usr/lib/ruby/vendor_ruby/rubygems/core_ext/kernel_require.rb:160:in require': cannot load such file -- robots (LoadError)

I think this means that the robots module is not available in ruby 2.7?

I just wanted to have the resolution for users affected here. The resolution is to install ruby 3, which is unfortunately not avialable for debian ?stable?. So I have this resolution to install ruby 3 with rvm. 

Could you restrict the version of ruby to 3? Or test the gem with ruby 2.7?

**Comment by buren** [comment]
> I think this means that the robots module is not available in ruby 2.7?

`robots` is a gem ([referenced here in wayback_archiver.gemspec](https://github.com/buren/wayback_archiver/blob/master/wayback_archiver.gemspec#L26))

It works for me on Ruby 2.7 👇 

<img width="839" alt="Screenshot 2022-01-21 at 09 45 06" src="https://user-images.githubusercontent.com/922411/150495249-a534b107-3dbc-4152-82be-2a85111b4a2c.png">

@chlorophyll-zz so there must be something else that is causing issues? 🤔

**Comment by chlorophyll-zz** [comment]
@buren Thank you. I am going to close this. I only have investigated this because #46  was mentioning that ruby version. I probably was running into that error when I tested with ruby 2.7. Btw can you close #46, this is completely resolved, but the user did not close the issue.

**Comment by buren** [comment]
Thanks!

**Comment by chlorophyll-zz** [comment]
I have this Issue again with ruby 3.0 in Ubuntu Jammy. I would prefer to not install ruby through rvm and keep the packaged ruby.

```
$ wayback_archiver 
<internal:/usr/lib/ruby/vendor_ruby/rubygems/core_ext/kernel_require.rb>:160:in `require': cannot load such file -- robots (LoadError)
	from <internal:/usr/lib/ruby/vendor_ruby/rubygems/core_ext/kernel_require.rb>:160:in `rescue in require'
	from <internal:/usr/lib/ruby/vendor_ruby/rubygems/core_ext/kernel_require.rb>:149:in `require'
	from /var/lib/gems/3.0.0/gems/wayback_archiver-1.4.0/lib/wayback_archiver/url_collector.rb:2:in `<top (required)>'
	from <internal:/usr/lib/ruby/vendor_ruby/rubygems/core_ext/kernel_require.rb>:85:in `require'
	from <internal:/usr/lib/ruby/vendor_ruby/rubygems/core_ext/kernel_require.rb>:85:in `require'
	from /var/lib/gems/3.0.0/gems/wayback_archiver-1.4.0/lib/wayback_archiver.rb:4:in `<top (required)>'
	from <internal:/usr/lib/ruby/vendor_ruby/rubygems/core_ext/kernel_require.rb>:85:in `require'
	from <internal:/usr/lib/ruby/vendor_ruby/rubygems/core_ext/kernel_require.rb>:85:in `require'
	from /var/lib/gems/3.0.0/gems/wayback_archiver-1.4.0/bin/wayback_archiver:4:in `<top (required)>'
	from /usr/local/bin/wayback_archiver:25:in `load'
	from /usr/local/bin/wayback_archiver:25:in `<main>'
<internal:/usr/lib/ruby/vendor_ruby/rubygems/core_ext/kernel_require.rb>:85:in `require': cannot load such file -- robots (LoadError)
	from <internal:/usr/lib/ruby/vendor_ruby/rubygems/core_ext/kernel_require.rb>:85:in `require'
	from /var/lib/gems/3.0.0/gems/wayback_archiver-1.4.0/lib/wayback_archiver/url_collector.rb:2:in `<top (required)>'
	from <internal:/usr/lib/ruby/vendor_ruby/rubygems/core_ext/kernel_require.rb>:85:in `require'
	from <internal:/usr/lib/ruby/vendor_ruby/rubygems/core_ext/kernel_require.rb>:85:in `require'
	from /var/lib/gems/3.0.0/gems/wayback_archiver-1.4.0/lib/wayback_archiver.rb:4:in `<top (required)>'
	from <internal:/usr/lib/ruby/vendor_ruby/rubygems/core_ext/kernel_require.rb>:85:in `require'
	from <internal:/usr/lib/ruby/vendor_ruby/rubygems/core_ext/kernel_require.rb>:85:in `require'
	from /var/lib/gems/3.0.0/gems/wayback_archiver-1.4.0/bin/wayback_archiver:4:in `<top (required)>'
	from /usr/local/bin/wayback_archiver:25:in `load'
	from /usr/local/bin/wayback_archiver:25:in `<main>'

```
```
gem list

*** LOCAL GEMS ***

abbrev (default: 0.1.0)
base64 (default: 0.1.0)
benchmark (default: 0.1.1)
bigdecimal (default: 3.0.0)
bundler (default: 2.2.22)
cgi (default: 0.2.0)
csv (default: 3.1.9)
date (default: 3.1.0)
dbm (default: 1.1.0)
debug (default: 0.1.0)
delegate (default: 0.2.0)
did_you_mean (default: 1.5.0)
digest (default: 3.0.0)
drb (default: 2.0.4)
english (default: 0.7.1)
erb (default: 2.2.0)
etc (default: 1.2.0)
fcntl (default: 1.0.0)
fiddle (default: 1.0.6)
fileutils (default: 1.5.0)
find (default: 0.1.0)
forwardable (default: 1.3.2)
gdbm (default: 2.1.0)
getoptlong (default: 0.1.1)
io-console (default: 0.5.7)
io-nonblock (default: 0.1.0)
io-wait (default: 0.1.0)
ipaddr (default: 1.2.2)
irb (default: 1.3.5)
json (default: 2.5.1)
logger (default: 1.4.3)
matrix (default: 0.3.1)
minitest (5.14.2)
mutex_m (default: 0.1.1)
net-ftp (default: 0.1.2)
net-http (default: 0.1.1)
net-imap (default: 0.1.1)
net-pop (default: 0.1.1)
net-protocol (default: 0.1.0)
net-smtp (default: 0.2.1)
net-telnet (0.1.1)
nkf (default: 0.1.0)
nokogiri (1.13.3 x86_64-linux)
observer (default: 0.1.1)
open-uri (default: 0.1.0)
open3 (default: 0.1.1)
openssl (default: 3.0.0)
optparse (default: 0.1.0)
ostruct (default: 0.3.1)
pathname (default: 0.1.0)
power_assert (1.2.0)
pp (default: 0.1.0)
prettyprint (default: 0.1.0)
prime (default: 0.1.2)
pstore (default: 0.1.1)
psych (default: 3.3.0)
racc (default: 1.5.1)
rake (13.0.6, 13.0.3)
rbs (1.0.4)
rdoc (default: 6.3.1)
readline (default: 0.0.2)
readline-ext (default: 0.1.1)
reline (default: 0.2.5)
resolv (default: 0.2.0)
resolv-replace (default: 0.1.0)
rexml (3.2.5)
rinda (default: 0.1.0)
robots (0.10.1)
rss (0.2.9)
rubygems-update (3.3.5)
securerandom (default: 0.1.0)
set (default: 1.0.1)
shellwords (default: 0.1.0)
singleton (default: 0.1.1)
spidr (0.6.1)
stringio (default: 3.0.0)
strscan (default: 3.0.0)
syslog (default: 0.1.0)
tempfile (default: 0.1.1)
test-unit (3.3.7)
time (default: 0.1.0)
timeout (default: 0.1.1)
tmpdir (default: 0.1.2)
tracer (default: 0.1.1)
tsort (default: 0.1.0)
typeprof (0.12.0)
un (default: 0.1.0)
uri (default: 0.10.1)
weakref (default: 0.1.1)
xmlrpc (0.3.2)
yaml (default: 0.1.1)
zlib (default: 1.1.0)

```

**Comment by bartman081523** [comment]
Issue is still pending, #64 is the fix

**Comment by dewey** [comment]
I think the best fix would be to vendor the one robots.rb file from the library as it seems the gem build process was just buggy, the code itself is fine. That's how metasploit did it: https://github.com/rapid7/metasploit-framework/commit/0e1bafb2d123019a42fe2524d897c0df2aeba4f8

**Comment by buren** [comment]
Thank you @dewey and everyone else for your patience. 

#72 follows @dewey suggestion. I've also updated the lib to properly support Ruby 3 and drop earlier EOL versions.

**Comment by buren** [comment]
[Published version 1.5.0 to RubyGems](https://rubygems.org/gems/wayback_archiver)
